File: solutions/leetcode/insert-delete-getrandom-o1/main.go
import "math/rand"

type RandomizedSet struct {
	Vals       []int
	ValToIndex map[int]int
}

func Constructor() RandomizedSet {
	return RandomizedSet{
		Vals:       []int{},
		ValToIndex: map[int]int{},
	}
}

func (this *RandomizedSet) Insert(val int) bool {
	if _, ok := this.ValToIndex[val]; ok {
		return false
	}

	this.ValToIndex[val] = len(this.Vals)
	this.Vals = append(this.Vals, val)

	return true
}

func (this *RandomizedSet) Remove(val int) bool {
	if _, ok := this.ValToIndex[val]; !ok {
		return false
	}

	valIndex := this.ValToIndex[val]
	lastVal := this.Vals[len(this.Vals)-1]

	this.ValToIndex[lastVal] = valIndex
	this.Vals[valIndex] = lastVal

	this.Vals = this.Vals[:len(this.Vals)-1]
	delete(this.ValToIndex, val)

	return true
}

func (this *RandomizedSet) GetRandom() int {
	return this.Vals[rand.Intn(len(this.Vals))]
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * obj := Constructor();
 * param_1 := obj.Insert(val);
 * param_2 := obj.Remove(val);
 * param_3 := obj.GetRandom();
 */

File: solutions/leetcode/insert-delete-getrandom-o1/main.py
class RandomizedSet:

    def __init__(self):
        self.val_index = dict()
        self.vals = list()

    def insert(self, val: int) -> bool:
        if val in self.val_index:
            return False
        self.val_index[val] = len(self.vals)
        self.vals.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self.val_index:
            return False

        idx = self.val_index[val]
        last_val = self.vals[-1]

        self.val_index[last_val] = idx
        self.vals[idx] = last_val

        self.vals.pop()
        del self.val_index[val]

        return True

    def getRandom(self) -> int:
        return random.choice(self.vals)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()
